### PR TITLE
Injector issues

### PR DESCRIPTION
**Pull request by ChristianBreitkreutz**
@dbucky It is better to do the discussion on the PR, it is more easy to point to the code

I am not sure if i find time today to have a look, but i will try

Update:

After all this back and forth. Ignore my last comments. I will do a example folder. And add an example how I see the the use-cases. This will take a while.  If you like you can later have a look if your Issues are covered.
Sorry for all this chaos

### CODE REVIEW

File: lib/Test/Mockify/Base.pm [hunk 1]
```diff
@@ -0,0 +1,249 @@
+package Test::Mockify::Base;
```
Comment (ChristianBreitkreutz) [inline lib/Test/Mockify/Base.pm:1]: Why you moved it to "base"? Please, move it back to mockify.pm 
Base is a PackageName which don't tell too much(base on base is a bit obvious ), and that the injector is basing on mockify is fine and tell the truth.

File: lib/Test/Mockify/Injector.pm [hunk 2]
```diff
@@ -0,0 +1,128 @@
+package Test::Mockify::Injector;
```
Comment (ChristianBreitkreutz) [inline lib/Test/Mockify/Injector.pm:1]: The name Injector is not a good name.  ( clean code rule: If you don't find a good name, is means there is a design issue) Nice article about this http://steve-yegge.blogspot.de/2006/03/execution-in-kingdom-of-nouns.html
I really think about this. The main idea is still super cool but.. What do you think about this?
It would be also in the chained style like the rest of mockify.
```
# The name sounds a bit funny but discribes best what this thing is doing
# -> It gives the posibility to create a scope in order to mock functions which are imported into your code.

my $MockScope = Test::Mockify::MockScope->new();
$MockScope->mockPackage('Package::Name')->mock('MethodName')->whenAny();
my $MockedPackage = $MockScope->mock('Package::Two');
$MockedPackage->mock('one')->when(String('hello'))->thenReturn('world');
$MockedPackage->spy('two')->whenAny();

# actual test    #############################################################
my $Sut = Sut->new();
$Sut->callWithStaticDependenies();
##############################################################################

$MockScope->clearPackage('Package::Name');
$MockScope->clearAll();
```

File: lib/Test/Mockify/Injector.pm [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    $mockedSelf->{'__MethodCallCounter'}->addMethod( $MethodName );
+    if(not $self->{'MethodStore'}{$MethodName}){
+        $self->{'MethodStore'}{$MethodName} //= $Method;
+        my $MockedMethodBody = sub {
+            my @MockedParameters = @_;
+            $mockedSelf->{'__MethodCallCounter'}->increment( $MethodName );
+            push @{$mockedSelf->{$MethodName.'_MockifyParams'}}, \@MockedParameters;
+            return $Method->call(@MockedParameters);
+        };
+        $self->{'override'}->replace($self->_mockedModulePath().'::'.$MethodName, $MockedMethodBody);
```
Comment (ChristianBreitkreutz) [inline lib/Test/Mockify/Injector.pm:36]: This covers the case of a full path call but not the imported version (I had to do it in my version )
```
use Path::To::Package qw (Function);
Function();
```
vs.
```
Path::To::Package::Function();
```

File: lib/Test/Mockify/Injector.pm [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+=pod
+
+=head1 NAME
+
+Test::Mockify::Injector - use L<Test::Mockify>'s robust API to override real functionality with mock implementations
+
+=head2 mock
+
+Sometimes it is not possible to inject test dependencies into your system under test (SUT).
+This could be the case for a variety of reasons, namely the SUT instantiates a new object instance, calls
```
Comment (ChristianBreitkreutz) [inline lib/Test/Mockify/Injector.pm:52]: > instantiates a new object instance

This is actually a design Issue in your code you like to test (could be solved with something like a factory-pattern), but i also understand that this is not very practical. I would keep it for now. This is  an Issue which should be solved later.

File: lib/Test/Mockify/Base.pm [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+    my ($MethodName) = @_;
+    my $PointerOriginalMethod = \&{$self->_mockedModulePath().'::'.$MethodName};
+    my $mockedSelf = $self->_mockedSelf();
+    #In order to have the current object available in the parameter list, it has to be injected here.
+    return $self->_addMockWithMethodSpy($MethodName, sub {
+            return $PointerOriginalMethod->($mockedSelf, @_);
+        });
+}
+
+sub getVerifier {
```
Comment (ChristianBreitkreutz) [inline lib/Test/Mockify/Base.pm:63]: You extend mockify so you should not change Mockify but your code ;-)

```
sub getVerifier {
  my $self = shift;
+    return $self->getMockObject();
+}
```
And, I am not 100% sure, but could it be that you break the scope of the injection when give out the object where the Sub::Override is created in? This was at least my problem in my solution.

File: t/FakeModuleStaticInjection.pm [hunk 6]
```diff
@@ -47,4 +49,12 @@ sub useImportedStaticFunction {
     return $PreFix . ': '.ReturnHelloWorld(@_);
 
 }
+sub dependency {
+    my ($self, $arg) = @_;
```
Comment (ChristianBreitkreutz) [inline t/FakeModuleStaticInjection.pm:56]: No constructor, No self! ;-)

File: t/Mockify_StaticInjection.t [hunk 7]
```diff
@@ -6,76 +6,87 @@ use lib ($FindBin::Bin);
...
 sub testPlan {
     my $self = shift;
 
     $self->test_mockStatic();
     $self->test_verify_with_mockAndSpy();
-    $self->test_functionNameFormatingErrorHandling();
     $self->test_MethodAndImportedFunctionHaveTheSameName();
-    $self->test_someSelectedMockifyFeatures();
-
+#    $self->test_someSelectedMockifyFeatures();
```
Comment (ChristianBreitkreutz) [inline t/Mockify_StaticInjection.t:29]: remove commented code ( clean code rule: Never commit commented code, if you need it for later use git diff)

File: t/Mockify_StaticInjection.t [hunk 8]
```diff
@@ -84,63 +95,145 @@ sub test_MethodAndImportedFunctionHaveTheSameName {
...
+    ok(WasCalled($injector->getVerifier(), 'breed'), "$SubTestName - Mocked breed method was called");
+    is(GetCallCount($injector->getVerifier(), 'breed'), 2, "$SubTestName - Called twice");
+}
+#----------------------------------------------------------------------------------------
+sub test_newNotCalled() {
+    my $self = shift;
+    my $SubTestName = (caller(0))[3];
+
+    my $mockify = Test::Mockify->new('Dog', ['Dalmatian']);
+    ok($mockify->getMockObject()->isa('Dog'), "$SubTestName - mock object isa Dog");
```
Comment (ChristianBreitkreutz) [inline t/Mockify_StaticInjection.t:221]: There is a isa assert
https://perldoc.perl.org/Test/More.html -> isa_ok

File: t/Mockify_StaticInjection.t [hunk 9]
```diff
@@ -84,63 +95,145 @@ sub test_MethodAndImportedFunctionHaveTheSameName {
...
+    my $originalValue = FakeModuleForMockifyTest::DummyMethodForTestOverriding();
+    my $originalValue2 = FakeModuleForMockifyTest::secondDummyMethodForTestOverriding();
+    my $mockValue = 'A mock dummy method';
+    my $mockValue2 = 'A second mock dummy method';
+    my $verifier;
+
+    {
+        my $injector = Test::Mockify::Injector->new('FakeModuleForMockifyTest');
+        $injector->mock('DummyMethodForTestOverriding')->when()->thenReturn($mockValue);
+        $injector->addMock('secondDummyMethodForTestOverriding', sub { $mockValue2 });
```
Comment (ChristianBreitkreutz) [inline t/Mockify_StaticInjection.t]: addMock is deprecated. Use  $injector->mock('secondDummyMethodForTestOverriding')->whenAny()->thenCall(sub {});
http://search.cpan.org/~chrisbr/Test-Mockify-1.3/lib/Test/Mockify/ReturnValue.pm#thenCall

Comment (ChristianBreitkreutz) [inline t/Mockify_StaticInjection.t]: when returning a value use return keyword. Yes, perl have an auto return but this is a hidden functionality and clean code should always be obvious.

**Review by ChristianBreitkreutz — COMMENTED** [review]
Hi Dustin,

sorry for the late answer. Unfortunately I had not  time last week.
I looking forward about your opinion on my suggestion ( comment in Injector.pm) If you think it is cool but you have no time, it is not a problem i can do this ( mostly ) during  work time.

Greetings,
Christian

File: lib/Test/Mockify/Base.pm (hunk 1 above)
Comment (dbucky) [reply lib/Test/Mockify/Base.pm:1]: I actually started with Injector inheriting from Mockify. This worked really well except for one thing: Mockify has the getMockObject method, but that name does not make sense for the Injector. The only reason I introduced Base was so Mockify could keep getMockObject, while Injector has getVerifier instead.

Maybe this could be resolved by removing the dual-purpose of Mockify's mock object, which serves as a test double and the verification object. Another idea I had could be part of the solution: perhaps the Mockify/Injector itself could be the verification object. There would be a getVerifier method, and additionally for Mockify, the getMockObject method. The inheritance would have to be switched to make that work (Mockify inherits from Injector), if you desire to remove Base.
```
Mockify
    getMockObject()

    ^

Injector
    getVerifier()
````
Base is meant to be abstract, such that it should not be instantiated itself.

File: lib/Test/Mockify/Injector.pm (hunk 2 above)
Comment (dbucky) [reply lib/Test/Mockify/Injector.pm:1]: This alternative doesn't provide much value that I can see beyond the current solution, or maybe I completely missed your point. MockScope seems to simply aggregate the various mocked packages into one place, though I don't see that being terribly useful or valuable. MockScope does provide explicit calls (clearPackage, clearAll) to restore the original functionality; however, if that is truly desired, we could easily add that to the existing implementation. Compare your implementation to what it would currently be:

```perl
my $packageNameInjector = Test::Mockify::Injector->new('Package::Name');
$packageNameInjector->mock('MethodName')->whenAny()->thenReturnUndef();
my $packageTwoInjector = Test::Mockify::Injector->new('Package::Two');
$packageTwoInjector->mock('one')->when(String('hello'))->thenReturn('world');
$packageTwoInjector->spy('two')->whenAny();

# actual test    #############################################################
my $Sut = Sut->new();
$Sut->callWithStaticDependenies();
##############################################################################
```

There is not a lot of difference. If you want the original implementation restored at the end, simply wrap that code in a block `{...}`, or we could add a `restore` function to the Injector so you can do that explicitly. I suppose if you're doing a bunch of mocking having the `clearAll` could be useful, but I doubt that will happen much.

```perl
$packageNameInjector->restore();
$packageTwoInjector->restore();
```

The current Injector implementation has just as much chaining capability as Mockify does. You could absolutely chain `Test::Mockify::Injector->new('Package::Name')->mock('MethodName')->whenAny()->thenReturn('ReturnValue')` if you don't care to have a reference to the Injector later (to mock/spy additional methods or for verification purposes). The same is true of Mockify.

Why don't you like Injector as a name? In the context of the way Mockify works and what the Injector is trying to change about that, I feel it adequately describes the functionality.

File: lib/Test/Mockify/Injector.pm (hunk 3 above)
Comment (dbucky) [reply lib/Test/Mockify/Injector.pm:36]: Why would you import a function into your package and then still call the fully qualified function name? I don't think this is a valid use case. Could you provide an example of why it would be?

If we assume I am correct in this thinking for a moment, you would simply mock the imported function via the package that imported it.

I get what you are saying about wanting to override the original package's implementation of a function, and all importing packages get that mocked functionality for free. My hesitation is that I don't think there is a way to achieve this without specifying both packages to the Injector, and that seems really clunky. That was essentially how you achieved it when you first overhauled my changes.

I think this is probably an edge case, anyway. Do you feel it is really important to address?

File: lib/Test/Mockify/Injector.pm (hunk 4 above)
Comment (dbucky) [reply lib/Test/Mockify/Injector.pm:52]: I'm not sure what you think needs to be solved here. I was merely stating that some application code is not written in a testable way, which is the exact problem the Injector was created to address: allowing you to test traditionally non-testable code.

File: t/Mockify_StaticInjection.t (hunk 7 above)
Comment (dbucky) [reply t/Mockify_StaticInjection.t:29]: I commented out some of the tests because they were breaking due to issues I had logged and intended to address. This is the main outstanding issue: https://github.com/dbucky/Mockify/issues/7.

File: t/Mockify_StaticInjection.t (hunk 8 above)
Comment (dbucky) [reply t/Mockify_StaticInjection.t:221]: I actually had forgotten that existed! I tried using it, but if I do, I cannot maintain the same test description format as the other tests. There is also not an equivalent for the negated form, which I am also trying to test just below that. However, feel free to change it to use isa_ok, if that's what you prefer.

File: t/Mockify_StaticInjection.t (hunk 9 above)
Comment (dbucky) [reply t/Mockify_StaticInjection.t]: I understand that addMock is deprecated. I intentionally included it in this test, because the Injector was not initially backwards-compatible with the deprecated methods. With my last round of changes, now it is. I suppose the Injector doesn't really need to be, and I see you created an Issue to remove them at some point. I was just trying to keep the Injector as close to Mockify as possible.

Comment (dbucky) [reply t/Mockify_StaticInjection.t]: Alright. I pushed an update for this!

File: t/FakeModuleStaticInjection.pm (hunk 6 above)
Comment (dbucky) [reply t/FakeModuleStaticInjection.pm:56]: I was following your lead for consistency with the rest of the file. But I agree, if the module does not represent a class, then there really shouldn't be any shifting for $self within each function. Furthermore, calls to that function should be of the form `FakeModuleStaticInjection::functionName`, not `FakeModuleStaticInjection->functionName`. I don't particularly care for this invocation flexibility in Perl. I will see about cleaning this up.

Comment (dbucky) [reply t/FakeModuleStaticInjection.pm:56]: I added some TODO statements as a reminder. I do not have time to get into it now.

File: lib/Test/Mockify/Base.pm (hunk 5 above)
Comment (dbucky) [reply lib/Test/Mockify/Base.pm:63]: As you probably read above, that function is mainly for the Injector. I'm not altogether happy with the way the getMockObject/getVerifier part turned out. You may have noticed that getMockObject simply calls getVerifier. At one point I had getMockObject and getVerifier on Mockify and Injector, respectively, and they each had the same implementation:

```perl
my $self = shift;
return $self->getMockObject();
```

That felt slightly wrong too, but to your point, it did not alter Mockify's implementation. I think if we work toward the solution I proposed to your comment about why did I create Base, this will be resolved.

File: lib/Test/Mockify/Injector.pm (hunk 2 above)
Comment (dbucky) [reply lib/Test/Mockify/Injector.pm:1]: In this case, the naming difficulty is less of a design issue in my opinion, and more the fact that it is difficult to capture in a succinct name the essence of what this module is trying to do.

File: lib/Test/Mockify/Base.pm (hunk 5 above)
Comment (dbucky) [reply lib/Test/Mockify/Base.pm:63]: The whole not going out of scope thing was because a direct reference to Mockify's $self was captured inside of several anonymous subs. This prevented the Mockify instance from ever being released. You can review this commit for how I solved it: https://github.com/dbucky/Mockify/commit/2924c49178c9d2594584ddb2ad25dea0368a27c5.

File: lib/Test/Mockify/Base.pm (hunk 1 above)
Comment (ChristianBreitkreutz) [reply lib/Test/Mockify/Base.pm:1]: You really made me start to think about this. But I have this gut feeling that it is not a good idea to have the base package. I think the actual issue is that the Injector is an additional Functionality for Mockify. This addition should not change the main project and it should be implemented with this in mind. I want it in a way that I could replace this addition without changing the original project. ( Your addition is still amazing but I want to keep the things separated because its keep the project clean)

You could solve the "getMockObject" vs. "getVerifier" Issue like this:
base:
```perl
package basePackage;

sub new {
    my $class = shift;
    return bless({}, $class);
}

sub baseMethod {
    my $self = shift;
    return (caller(0))[3];
}

1;
```
child
```perl
package childPackage;
use base qw(basePackage);

sub new {
    my $class = shift;
    my $self = $class->SUPER::new();
    return bless({}, $class);
}

sub baseMethod {
    my $self = shift;
    die("'Please use 'sameButNotBaseMethod' instead of 'baseMethod'");
    return;
}
sub sameButNotBaseMethod {
    my $self = shift;
    return $self->SUPER::baseMethod();
}

1;
```
test ( use cases )
```perl
package baseTest;
use strict;
use FindBin;
use lib ($FindBin::Bin);
use parent 'TestBase';
use Test::More;

use basePackage;
use childPackage;
use Test::Exception;
#------------------------------------------------------------------------
sub testPlan{
    my $self = shift;
    my $BasePackage = basePackage->new();
    is($BasePackage->baseMethod(),'basePackage::baseMethod', 'prove call of base package method' );

    my $ChildPackage = childPackage->new();
    is($ChildPackage->sameButNotBaseMethod(),'basePackage::baseMethod', 'prove piped call to base package method' );

    throws_ok( sub { $ChildPackage->baseMethod() },
                   qr/Please use 'sameButNotBaseMethod' instead of 'baseMethod'/,
                   " prove the expected Error when sombody try to call the baseMethod of the child"
         );
    return;
}

__PACKAGE__->RunTest();
1;
```

File: lib/Test/Mockify/Injector.pm (hunk 2 above)
Comment (ChristianBreitkreutz) [reply lib/Test/Mockify/Injector.pm:1]: Yes, you are right. I didn't saw this possibility of chaining and this is absolutely enough. 
```perl
Test::Mockify::Injector->new('Package::Name')->mock('MethodName')->whenAny()->thenReturn('ReturnValue')
```

Comment (ChristianBreitkreutz) [reply lib/Test/Mockify/Injector.pm:1]: So I would mainly keep what you did (if the base thing is solved), but the name is not acceptable for me. "Injection" is a verb. "Injector" is a nominalized verb of it.
But let me share my thoughts:
I think we agree on that the Injector gives the possibility to inject( and here is the verb correct) a mock into something.  The Question is what is something, so where we inject the mock. Inject something into the Injector make no sense for my point of view. I think we **inject** a **mock of a function** into a **Package** which is valid in our **current scope**. So i would suggest "Package" instead of "Injector". And even one step further we could rename the action(verb) from "mock" to "mockFunction"
```perl
{ # this defines the scope
    my PackageTool = Test::Mockify::Package->new('Some::Tool');
    PackageTooll->mockFunction('Greet')->when(String('Dustin'))->thenReturn('Hello Dustin');
    PackageTool->mockFunction('Greet')->when(String('Christian'))->thenReturn('Hello Christian');
    PackageTool->mockFunction('SayBye')->whenAny()->thenReturn('Bye All');
}
```  
what do you think?

File: lib/Test/Mockify/Injector.pm (hunk 3 above)
Comment (ChristianBreitkreutz) [reply lib/Test/Mockify/Injector.pm:36]: I have the impression we did't talk about the same thing. I will have now a look again into the PR. Lets see what I misunderstood.

Comment (ChristianBreitkreutz) [reply lib/Test/Mockify/Injector.pm:36]: ok, I think now i get it. The thing I didn't realized was this:
```perl
my $injectorImported = Test::Mockify::Injector->new('FakeModuleStaticInjection');
    $injectorImported->mock('ReturnHelloWorld')->when(String('German'))->thenReturn('Hallo Welt');
```
I didn't realize that the injector doesn't create a 'FakeStaticTools' but a 'FakeModuleStaticInjection'. And than there is mock a for a Function which is not owned by this Package. This sounds wrong to me. Off course when using "use name qw (Function)" this Function will become part of this module and the "module->can()" check is true. But this is in-transparent. In my company we are using only this import option. This would result in unmaintainable test code. 
For the full qualified Call your solution is fine but not for imported functions.
I think this is the reason for the two different views on this injecting problem, namings, aso.  I will rethink the whole problem. and let you know.

File: lib/Test/Mockify/Base.pm (hunk 5 above)
Comment (ChristianBreitkreutz) [reply lib/Test/Mockify/Base.pm:63]: wow! this fixes my scope problems. THX I didn't found it.

**Comment by ChristianBreitkreutz** [comment]
After some consideration and arguing I decided for fix this differently. I will keep mentioning and smaller code changes from dbucky.  See https://github.com/ePages-de/Mockify/pull/33
